python3Packages.vllm: disable Blackwell GPU support to fix CUDA build

daniel-fahey opened the pull request:
Fixes type of build failure seen [here](https://hydra.nixos-cuda.org/build/824/nixlog/71).
- vLLM v0.11.0's FlashMLA kernels for Blackwell (SM100+) GPUs require CUTLASS v4.2.1+ APIs that are not available in the bundled v4.0.0
- Temporarily disable these architectures until upstream upgrades CUTLASS

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage reviewed (commented):
Have you tried bumping `cutlass` to `4.2.1` instead?

daniel-fahey commented:
> Have you tried bumping `cutlass` to `4.2.1` instead?

Not yet, definitely worth a go.

daniel-fahey commented on pkgs/development/python-modules/vllm/default.nix:
```diff
@@ -438,6 +438,11 @@ buildPythonPackage rec {
     (lib.cmakeFeature "CUTLASS_ENABLE_CUBLAS" "ON")
+    # Disable SM100 kernels - requires CUTLASS v4.2.1+ but v0.11.0 uses v4.0.0
+    # See: https://github.com/vllm-project/vllm/pull/24673
+    (lib.cmakeFeature "ENABLE_SCALED_MM_SM100" "OFF")
+    (lib.cmakeFeature "ENABLE_NVFP4_SM100" "OFF")
+    (lib.cmakeFeature "ENABLE_CUTLASS_MOE_SM100" "OFF")
```
Nitpicks welcome. Probably, and looks more idiomatic. Not sure this approach will even work, though, lol. I'm still at the "Make it Work" stage.

daniel-fahey commented:
```console
nix-build -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/0da19742a9d0b7041c791cb043ffaaccffb6090d.tar.gz --arg config '{ allowUnfree = true; cudaSupport = true; }' -A python313Packages.vllm
```
failing 
```console
[nixbuild.net] Build 4586852 finished in 27 minutes with status 'build_failed'. Max memory usage was 73.0 of 128.5 GB. Average CPU utilization was 84% of 15 CPUs.
```
[See this link for more details and logs](https://nixbuild.net/builds/4586852?t=EtkBCm8KBWJ1aWxkCgpidWlsZDpyZWFkGAMiCQoHCAcSAxD5BSINCgsIBBIHOgUKAxiBCDImCiQKAggbEgYIBRICCAUaFgoECgIIBQoICgYgrLu80wYKBBoCCAAyFgoUCgIIGxIOCAISAxiACBIFEOT6lwISJAgAEiBiVTrcM8boE4msfiMtY6r30zhRtQdwII-OPjOd5Hj84RpABJIwJr4GBl4iqi7XyjMQXgxC3LT_PlNyef1rOG-dk1AKRtRDLAjSBMl7wSwm-gX7B4ofXTrtq7hWsy0ZkWe7CSIiCiAtp0eM-2aX23SEokwGeOE59QJ2kIT2DHvVXbRr2azZ_A==).

daniel-fahey commented:
c25087d:
```bash
nix-build --pure https://github.com/NixOS/nixpkgs/archive/c25087d4ad1c47ed6515521b3600c8e32938f596.tar.gz --arg config '{ allowUnfree = true; cudaSupport = true; }' --attr python313Packages.vllm
```
```console
this derivation will be built:
  /nix/store/7f5rgisqx4k925a2hqpf6s3qf0hsc5kz-python3.13-vllm-0.11.0.drv
building '/nix/store/7f5rgisqx4k925a2hqpf6s3qf0hsc5kz-python3.13-vllm-0.11.0.drv' on 'ssh://eu.nixbuild.net'...
copying 0 paths...
[nixbuild.net] Build 4590083 queued at 25-10-06 15:30:00Z. Allocated 15 CPUs, 128.5 GB memory.
```
(will share results when it's completed)

Commit:
python3Packages.vllm: disable Blackwell GPU support to fix CUDA build

- vLLM v0.11.0's FlashMLA kernels for Blackwell (SM100+) GPUs require CUTLASS v4.2.1+ APIs that are not available in the bundled v4.0.0
- Temporarily disable these architectures until upstream upgrades CUTLASS

GaetanLepage reviewed (approved):
Thanks
